src/libDirectoryService/PoWProcessing.cpp

```diff
@@ -38,149 +39,70 @@
+                cv_lk, std::chrono::seconds(POW_SUBMISSION_TIMEOUT))
+            == std::cv_status::timeout)
+        {
+            LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
+                      "Time out while waiting for state transition ");
+        }
 
-    // Non-gensis block
-    if (block_num > 1)
-    {
-        expectedDSDiff = m_mediator.m_dsBlockChain.GetLastBlock()
-                             .GetHeader()
-                             .GetDSDifficulty();
-        expectedDiff = m_mediator.m_dsBlockChain.GetLastBlock()
-                           .GetHeader()
-                           .GetDifficulty();
+        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
+                  "State transition is completed. (check for timeout)");
     }
 
-    if (difficultyLevel != expectedDSDiff && difficultyLevel != expectedDiff)
+    if (!CheckState(PROCESS_POWSUBMISSION))
     {
-        LOG_GENERAL(WARNING,
-                    "Difficulty level is invalid. difficultyLevel: "
-                        << to_string(difficultyLevel)
-                        << " Expected: " << to_string(expectedDSDiff) << " or "
-                        << to_string(expectedDiff));
-
-        // TODO: penalise sender in reputation manager
+        LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
+                  "Not at POW_SUBMISSION. Current state is " << m_state);
         return false;
     }
 
-    m_timespec = r_timer_start();
-    bool result = POW::GetInstance().PoWVerify(
-        block_num, difficultyLevel, rand1, rand2, from.m_ipAddress, key, false,
-        nonce, winning_hash, winning_mixhash);
-    LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-              "[POWSTAT] pow verify (microsec): " << r_timer_end(m_timespec));
-
-    return result;
-}
-
-bool DirectoryService::ParseMessageAndVerifyPOW(
-    const vector<unsigned char>& message, unsigned int offset, const Peer& from)
-{
-    if (LOOKUP_NODE_MODE)
-    {
-        LOG_GENERAL(WARNING,
-                    "DirectoryService::ParseMessageAndVerifyPOW not expected "
-                    "to be called from LookUp node.");
-        return true;
-    }
-
-    unsigned int curr_offset = offset;
-    // 8-bytes block number
-    uint64_t DSBlockNum = Serializable::GetNumber<uint64_t>(
-        message, curr_offset, sizeof(uint64_t));
-    curr_offset += sizeof(uint64_t);
-
-    // 1-byte difficultyLevel
-    uint8_t difficultyLevel = Serializable::GetNumber<uint8_t>(
-        message, curr_offset, sizeof(uint8_t));
-    curr_offset += sizeof(uint8_t);
-
-    // Check block number
-    if (!CheckWhetherDSBlockIsFresh(DSBlockNum))
+    uint64_t blockNumber = 0;
+    uint8_t difficultyLevel = 0;
+    Peer submitterPeer;
+    PubKey submitterPubKey;
+    uint64_t nonce = 0;
+    string resultingHash;
+    string mixHash;
+    Signature signature;
+
+    if (!Messenger::GetDSPoWSubmission(
```

Comment:
GetDSPoWSubmission can be misleading. This can be pow submission using diff or dsdiff. 

The node declare either it is using diff or dsdiff in PoW submission. DS commitee check the declared diff is strictly either ds diff or diff and proceeds to verify using dsdiff or diff.

Reply:
To summarize our discussion just now: The DS in the function name is based on convention (DS meaning a message meant for DS; other messages will be pre-pended with Node or Lookup). So it doesn't actually mean DS-level PoW submission.